cms/djangoapps/contentstore/rest_api/v2/views/downstreams.py

```diff
@@ -0,0 +1,227 @@
+    #     downstream = _load_accessible_block(request.user, usage_key_string, require_write_access=True)
+    #     _ensure_upstream_ref(downstream)
+    #     ....
+
+
+@view_auth_classes(is_authenticated=True)
+class SyncFromUpstreamView(DeveloperErrorViewMixin, APIView):
+    """
+    Accept or decline an opportunity to sync a downstream block from its upstream content.
+    """
+
+    def post(self, request: Request, usage_key_string: str) -> Response:
+        """
+        Pull latest updates to the block at {usage_key_string} from its linked upstream content.
+        """
+        assert isinstance(request.user, AbstractUser)
+        downstream = _load_accessible_block(request.user, usage_key_string, require_write_access=True)
+        _ensure_upstream_ref(downstream)
+        if not downstream.upstream:
+            raise NotFound(detail=f"Block '{usage_key_string}' is not linked to a library block")
+        old_version = downstream.upstream_version
+        try:
+            sync_from_upstream(downstream, request.user)
+        except (BadUpstream, BadDownstream) as exc:
+            raise ValidationError(detail=str(exc)) from exc
+        modulestore().update_item(downstream, request.user.id)
+        upstream_link = UpstreamLink.get_for_block(downstream)
+        assert upstream_link
+        return Response(upstream_link.to_json(), status=200)
+
+    def delete(self, request: Request, usage_key_string: str) -> Response:
+        """
+        Decline the latest updates to the block at {usage_key_string}.
+        """
+        assert isinstance(request.user, AbstractUser)
+        downstream = _load_accessible_block(request.user, usage_key_string, require_write_access=True)
+        _ensure_upstream_ref(downstream)
+        try:
+            decline_sync(downstream)
+        except (BadUpstream, BadDownstream) as exc:
+            raise ValidationError(str(exc)) from exc
+        modulestore().update_item(downstream, request.user.id)
+        return Response(status=204)
+
+
+def _load_accessible_block(user: AbstractUser, usage_key_string: str, *, require_write_access: bool) -> XBlock:
+    """
+    Given a logged in-user and a serialized usage key of an upstream-linked XBlock, load it from the ModuleStore,
+    raising a DRF-friendly exception if anything goes wrong.
+
+    Raises NotFound if usage key is malformed.
+    Raises NotFound if user lacks access.
+    Raises NotFound if block does not exist.
+    """
+    not_found = NotFound(detail=f"Block not found or not accessible: {usage_key_string}")
+    try:
+        usage_key = UsageKey.from_string(usage_key_string)
+    except InvalidKeyError as exc:
+        raise ValidationError(detail=f"Malformed block usage key: {usage_key_string}") from exc
+    if require_write_access and not has_studio_write_access(user, usage_key.context_key):
+        raise not_found
+    if not has_studio_read_access(user, usage_key.context_key):
+        raise not_found
```

Comment:
Should this log an error as well?

Reply:
@ormsbee AFAICT, we don't typically do additional logging for cut-and-dry 403/404 scenarios, so I'm not sure why we would here.

Reply:
Fair point. The main reason I bring it up is because in this case, it might be confusing to see a bunch of 404s in the logs for this, when it's really an underlying authz check that's failing.

Reply:
(Not asking for changes now.) Personally, I've really come to appreciate a clear separation between 403 and 404 errors for things like this. Makes debugging, messaging to users, and supporting users much easier. "Leaking" what IDs exist or do not exist in a course is not really a big security concern, and would only apply to users who somehow have permission to read some parts of the course but not other parts - a tiny subset of user requests.

Reply:
@bradenmacdonald I do like to use the full range of expressiveness that HTTP gives us, for all the reasons you highlight. So I'm totally open to changing this to 403 if we deem it not to be an integrity issue. I'm leaving it for now, but I'll add it to my list of things to follow up on.